File: packages/server/src/shared-tests.ts
/**
 * SHARED store test suite.
 *
 * Every `Store` implementation must pass this exact matrix — the same
 * idea as the client's shared adapter suite
 * (`@vidcall/transport/shared-tests`). Run it from each store's test file:
 *
 * ```ts
 * import { runStoreTestSuite } from '@vidcall/server/shared-tests';
 * runStoreTestSuite({
 *   name: 'sqlite',
 *   createStore: async () => { const s = new SqliteStore(new Database(':memory:')); await s.bootstrap(); return s; },
 * });
 * ```
 */

import { beforeEach, afterEach, describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { createEnvelope } from '@vidcall/protocol';
import {
  closeRoom,
  createRoom,
  getRoomState,
  handleSignal,
  joinRoom,
  leaveRoom,
  startRecording,
  stopRecording,
} from './core.js';
import type { Store } from './store.js';
import type { ParticipantInput } from './core.js';

export interface StoreHarness {
  /** Store name (describe title). */
  name: string;
  /** Fresh, empty store per test. */
  createStore(): Promise<Store>;
  /** Tear down (close pools, delete temp dirs, ...). */
  destroyStore?(store: Store): Promise<void>;
  /** Unique-room-id prefix to avoid cross-run collisions. Default 'shared'. */
  roomPrefix?: string;
}

let suiteCounter = 0;

export function runStoreTestSuite(h: StoreHarness): void {
  describe(`@vidcall/server shared store suite → ${h.name}`, () => {
    let store: Store;
    beforeEach(async () => {
      store = await h.createStore();
    });
    afterEach(async () => {
      await h.destroyStore?.(store);
    });

    function roomId(): string {
      return `${h.roomPrefix ?? 'shared'}-${h.name}-r${++suiteCounter}`;
    }

    function participant(id: string): ParticipantInput {
      return { participantId: id, sessionId: `session-${id}`, displayName: `User ${id}` };
    }

    it('creates a room and reads it back with metadata intact', async () => {
      const id = roomId();
      const room = await createRoom(store, {
        roomId: id,
        maxParticipants: 4,
        metadata: { topic: 'standup', nested: { a: [1, 2, 3] } },
      });
      const fetched = await store.getRoom(id);
      assert.ok(fetched);
      assert.equal(fetched.roomId, id);
      assert.equal(fetched.state, 'open');
      assert.equal(fetched.maxParticipants, 4);
      assert.deepEqual(fetched.metadata, { topic: 'standup', nested: { a: [1, 2, 3] } });
      assert.equal(room.createdAt, fetched.createdAt);
    });

    it('rejects a duplicate room id', async () => {
      const id = roomId();
      await createRoom(store, { roomId: id });
      await assert.rejects(createRoom(store, { roomId: id }), (err: unknown) =>
        (err as { code?: string }).code === 'room_already_exists');
    });

    it('join adds a participant and listParticipants reflects the roster', async () => {
      const id = roomId();
      await createRoom(store, { roomId: id });
      const a = await joinRoom(store, id, participant('alice'));
      const b = await joinRoom(store, id, participant('bob'));
      assert.equal(a.participants.length, 1);
      assert.equal(b.participants.length, 2);
      const roster = await store.listParticipants(id);
      assert.deepEqual(roster.map((p) => p.participantId), ['alice', 'bob']);
      const alice = await store.getParticipant(id, 'alice');
      assert.equal(alice?.displayName, 'User alice');
      assert.equal(alice?.sessionId, 'session-alice');
      assert.equal(alice?.joinedAt, a.participant.joinedAt);
    });

    it('join on an unknown room fails with room_not_found', async () => {
      await assert.rejects(joinRoom(store, roomId(), participant('alice')), (err: unknown) =>
        (err as { code?: string }).code === 'room_not_found');
    });

    it('double-join of the same participant fails unless upsert', async () => {
      const id = roomId();
      await createRoom(store, { roomId: id });
      await joinRoom(store, id, participant('alice'));
      await assert.rejects(joinRoom(store, id, participant('alice')), (err: unknown) =>
        (err as { code?: string }).code === 'participant_already_joined');
      const result = await joinRoom(store, id, participant('alice'), { upsert: true });
      assert.equal(result.participants.length, 1);
    });

    it('enforces maxParticipants', async () => {
      const id = roomId();
      await createRoom(store, { roomId: id, maxParticipants: 1 });
      await joinRoom(store, id, participant('alice'));
      await assert.rejects(joinRoom(store, id, participant('bob')), (err: unknown) =>
        (err as { code?: string }).code === 'room_full');
    });

    it('closeRoom rejects joins but keeps existing members signaling', async () => {
      const id = roomId();
      await createRoom(store, { roomId: id });
      await joinRoom(store, id, participant('alice'));
      const closed = await closeRoom(store, id);
      assert.equal(closed.state, 'closed');
      await assert.rejects(joinRoom(store, id, participant('bob')), (err: unknown) =>
        (err as { code?: string }).code === 'room_closed');
      const delivery = await handleSignal(
        store,
        createEnvelope('chat', {
          roomId: id,
          senderId: 'alice',
          sessionId: 'session-alice',
          payload: { text: 'still here' },
        }),
      );
      assert.equal(delivery.recipients.length, 0); // alice is the only member
    });

    it('leave removes the participant and leaves are idempotent-safe (second leave errors)', async () => {
      const id = roomId();
      await createRoom(store, { roomId: id });
      await joinRoom(store, id, participant('alice'));
      await joinRoom(store, id, participant('bob'));
      const result = await leaveRoom(store, id, 'alice');
      assert.deepEqual(result.participants.map((p) => p.participantId), ['bob']);
      assert.equal(await store.getParticipant(id, 'alice'), null);
      await assert.rejects(leaveRoom(store, id, 'alice'), (err: unknown) =>
        (err as { code?: string }).code === 'participant_not_found');
    });

    it('putSignal assigns strictly increasing per-room seq; listSignals(since) filters', async () => {
      const id = roomId();
      await createRoom(store, { roomId: id });
      await joinRoom(store, id, participant('alice'));
      const s1 = await store.putSignal({
        roomId: id,
        envelope: createEnvelope('chat', { roomId: id, senderId: 'alice', sessionId: 's', payload: { text: 'one' } }),
        receivedAt: 1,
      });
      const s2 = await store.putSignal({
        roomId: id,
        envelope: createEnvelope('chat', { roomId: id, senderId: 'alice', sessionId: 's', payload: { text: 'two' } }),
        receivedAt: 2,
      });
      const s3 = await store.putSignal({
        roomId: id,
        envelope: createEnvelope('chat', { roomId: id, senderId: 'alice', sessionId: 's', payload: { text: 'three' } }),
        receivedAt: 3,
      });
      assert.ok(s1.seq < s2.seq && s2.seq < s3.seq);
      const after = await store.listSignals(id, s2.seq);
      assert.deepEqual(after.map((s) => s.seq), [s3.seq]);
      const all = await store.listSignals(id, 0);
      assert.equal(all.length, 3);
      // Envelope JSON round-trips verbatim through the store.
      assert.deepEqual(all[1]!.envelope, s2.envelope);
    });

    it('handles relay recipient rules: offer excludes sender, presence includes everyone', async () => {
      const id = roomId();
      await createRoom(store, { roomId: id });
      await joinRoom(store, id, participant('alice'));
      await joinRoom(store, id, participant('bob'));
      await joinRoom(store, id, participant('carol'));

      const offer = await handleSignal(
        store,
        createEnvelope('offer', {
          roomId: id,
          senderId: 'alice',
          sessionId: 'session-alice',
          targetSenderId: 'carol',
          payload: { sdp: 'v=0\r\n' },
        }),
      );
      assert.deepEqual(offer.recipients.map((p) => p.participantId), ['carol']);

      const broadcast = await handleSignal(
        store,
        createEnvelope('presence', {
          roomId: id,
          senderId: 'bob',
          sessionId: 'session-bob',
          payload: { state: 'online' },
        }),
      );
      assert.deepEqual(broadcast.recipients.map((p) => p.participantId).sort(), ['alice', 'bob', 'carol']);
    });

    it('rejects signals from participants who are not in the room', async () => {
      const id = roomId();
      await createRoom(store, { roomId: id });
      await joinRoom(store, id, participant('alice'));
      await assert.rejects(
        handleSignal(
          store,
          createEnvelope('chat', { roomId: id, senderId: 'eve', sessionId: 'session-eve', payload: { text: 'hi' } }),
        ),
        (err: unknown) => (err as { code?: string }).code === 'participant_not_found',
      );
    });

    it('rejects malformed envelopes', async () => {
      const id = roomId();
      await createRoom(store, { roomId: id });
      await assert.rejects(handleSignal(store, { not: 'an envelope' }), (err: unknown) =>
        (err as { code?: string }).code === 'invalid_envelope');
    });

    it('touches lastSeenAt on activity', async () => {
      const id = roomId();
      await createRoom(store, { roomId: id });
      const { participant: alice } = await joinRoom(store, id, participant('alice'));
      await handleSignal(
        store,
        createEnvelope('reaction', { roomId: id, senderId: 'alice', sessionId: 'session-alice', payload: { emoji: '👋' } }),
      );
      const updated = await store.getParticipant(id, 'alice');
      assert.ok(updated!.lastSeenAt >= alice.lastSeenAt);
    });

    it('recording sessions round-trip through the store', async () => {
      const id = roomId();
      await createRoom(store, { roomId: id });
      const rec = await startRecording(store, id, { metadata: { mime: 'video/webm' } });
      assert.equal(rec.status, 'recording');
      assert.equal(rec.roomId, id);
      const list = await store.listRecordings(id);
      assert.equal(list.length, 1);
      assert.deepEqual(list[0]!.metadata, { mime: 'video/webm' });
      const fetched = await store.getRecording(rec.sessionId);
      assert.equal(fetched?.sessionId, rec.sessionId);
      const stopped = await stopRecording(store, rec.sessionId);
      assert.equal(stopped.status, 'finalized');
      assert.ok(stopped.stoppedAt! >= stopped.startedAt);
      const byId = await store.getRecording(rec.sessionId);
      assert.equal(byId?.status, 'finalized');
      await assert.rejects(stopRecording(store, 'missing'), (err: unknown) =>
        (err as { code?: string }).code === 'recording_not_found');
    });

    it('getRoomState returns room + roster + signal count', async () => {
      const id = roomId();
      await createRoom(store, { roomId: id });
      await joinRoom(store, id, participant('alice'));
      await handleSignal(
        store,
        createEnvelope('chat', { roomId: id, senderId: 'alice', sessionId: 'session-alice', payload: { text: 'x' } }),
      );
      const state = await getRoomState(store, id);
      assert.equal(state.room.roomId, id);
      assert.equal(state.participants.length, 1);
      assert.ok(state.signalCount >= 1);
    });

    it('deleteRoom clears participants, signals and recordings (when supported)', async () => {
      const id = roomId();
      await createRoom(store, { roomId: id });
      await joinRoom(store, id, participant('alice'));
      const rec = await startRecording(store, id);
      await store.putSignal({
        roomId: id,
        envelope: createEnvelope('chat', { roomId: id, senderId: 'alice', sessionId: 'session-alice', payload: { text: 'x' } }),
        receivedAt: 1,
      });
      if (!store.deleteRoom) return;
      await store.deleteRoom(id);
      assert.equal(await store.getRoom(id), null);
      assert.equal((await store.listParticipants(id)).length, 0);
      assert.equal((await store.listSignals(id, 0)).length, 0);
      assert.equal((await store.listRecordings(id)).length, 0);
      assert.equal(await store.getRecording(rec.sessionId), null);
    });
  });
}

File: server/rust/crates/vidcall-server/src/shared_tests.rs
//! SHARED store test suite — mirror of `packages/server/src/shared-tests.ts`.
//!
//! Every [`Store`] implementation must pass this exact matrix — the same idea
//! as the client's shared adapter suite (`@vidcall/transport/shared-tests`).
//! Run it from each store's test file:
//!
//! ```
//! use vidcall_server::stores::InMemoryStore;
//! use vidcall_server::shared_tests::{run_store_test_suite, StoreHarness};
//!
//! #[tokio::test]
//! async fn memory_passes_shared_suite() {
//!     run_store_test_suite(StoreHarness {
//!         name: "memory",
//!         create_store: Box::new(|| Box::pin(async { Ok(Box::new(InMemoryStore::new()) as Box<dyn Store>) })),
//!         destroy_store: None,
//!         room_prefix: Some("shared".to_string()),
//!     }).await;
//! }
//! ```
//!
//! The suite runs inside a single `tokio::test` so stores that need one
//! runtime (e.g. a background poller) are easy to drive.

use std::sync::atomic::{AtomicU64, Ordering};

use serde_json::json;

use crate::core::{
    close_room, create_room, get_room_state, handle_signal, join_room, leave_room, start_recording,
    stop_recording, ParticipantInput,
};
use crate::error::VidcallError;
use crate::protocol::create_envelope;
use crate::store::Store;

static SUITE_COUNTER: AtomicU64 = AtomicU64::new(0);

/// Fresh-store factory used by the shared suite.
pub struct StoreHarness {
    /// Store name (used in failure messages / unique room ids).
    pub name: &'static str,
    /// Fresh, empty store per test.
    pub create_store: Box<
        dyn Fn() -> std::pin::Pin<
                Box<dyn std::future::Future<Output = crate::error::Result<Box<dyn Store>>> + Send>,
            > + Send
            + Sync,
    >,
    /// Tear down (close pools, delete temp dirs, ...).
    pub destroy_store: Option<
        Box<
            dyn Fn(
                    Box<dyn Store>,
                )
                    -> std::pin::Pin<Box<dyn std::future::Future<Output = ()> + Send>>
                + Send
                + Sync,
        >,
    >,
    /// Unique-room-id prefix to avoid cross-run collisions.
    pub room_prefix: Option<String>,
}

fn room_id(prefix: &str, name: &str) -> String {
    let n = SUITE_COUNTER.fetch_add(1, Ordering::SeqCst);
    format!("{prefix}-{name}-r{n}")
}

fn participant(id: &str) -> ParticipantInput {
    ParticipantInput {
        participant_id: id.to_string(),
        session_id: format!("session-{id}"),
        display_name: Some(format!("User {id}")),
        metadata: None,
    }
}

/// Run the full shared suite against one store.
pub async fn run_store_test_suite(h: StoreHarness) {
    let prefix = h
        .room_prefix
        .clone()
        .unwrap_or_else(|| "shared".to_string());
    let name = h.name;

    // ---- rooms -----------------------------------------------------------
    {
        let store = fresh(&h).await;
        let id = room_id(&prefix, name);
        let room = create_room(
            &*store,
            crate::core::CreateRoomOptions {
                room_id: Some(id.clone()),
                max_participants: Some(4),
                metadata: Some(json!({"topic": "standup", "nested": {"a": [1, 2, 3]}})),
                now: None,
            },
        )
        .await
        .expect("create_room");
        let fetched = store.get_room(&id).await.unwrap().expect("room exists");
        assert_eq!(fetched.room_id, id);
        assert_eq!(format!("{:?}", fetched.state), "Open");
        assert_eq!(fetched.max_participants, Some(4));
        assert_eq!(
            fetched.metadata,
            Some(json!({"topic": "standup", "nested": {"a": [1, 2, 3]}}))
        );
        assert_eq!(room.created_at, fetched.created_at);
        destroy(&h, store).await;
    }

    {
        let store = fresh(&h).await;
        let id = room_id(&prefix, name);
        create_room(
            &*store,
            crate::core::CreateRoomOptions {
                room_id: Some(id.clone()),
                ..Default::default()
            },
        )
        .await
        .unwrap();
        let err = create_room(
            &*store,
            crate::core::CreateRoomOptions {
                room_id: Some(id.clone()),
                ..Default::default()
            },
        )
        .await
        .unwrap_err();
        assert_eq!(err.code, "room_already_exists");
        destroy(&h, store).await;
    }

    // ---- join / roster ---------------------------------------------------
    {
        let store = fresh(&h).await;
        let id = room_id(&prefix, name);
        create_room(
            &*store,
            crate::core::CreateRoomOptions {
                room_id: Some(id.clone()),
                ..Default::default()
            },
        )
        .await
        .unwrap();
        let a = join_room(&*store, &id, participant("alice"), Default::default())
            .await
            .unwrap();
        let b = join_room(&*store, &id, participant("bob"), Default::default())
            .await
            .unwrap();
        assert_eq!(a.participants.len(), 1);
        assert_eq!(b.participants.len(), 2);
        let roster = store.list_participants(&id).await.unwrap();
        assert_eq!(
            roster
                .iter()
                .map(|p| p.participant_id.as_str())
                .collect::<Vec<_>>(),
            vec!["alice", "bob"]
        );
        let alice = store
            .get_participant(&id, "alice")
            .await
            .unwrap()
            .expect("alice");
        assert_eq!(alice.display_name.as_deref(), Some("User alice"));
        assert_eq!(alice.session_id, "session-alice");
        assert_eq!(alice.joined_at, a.participant.joined_at);
        destroy(&h, store).await;
    }

    {
        let store = fresh(&h).await;
        let err = join_room(
            &*store,
            &room_id(&prefix, name),
            participant("alice"),
            Default::default(),
        )
        .await
        .unwrap_err();
        assert_eq!(err.code, "room_not_found");
        destroy(&h, store).await;
    }

    {
        let store = fresh(&h).await;
        let id = room_id(&prefix, name);
        create_room(
            &*store,
            crate::core::CreateRoomOptions {
                room_id: Some(id.clone()),
                ..Default::default()
            },
        )
        .await
        .unwrap();
        join_room(&*store, &id, participant("alice"), Default::default())
            .await
            .unwrap();
        let err = join_room(&*store, &id, participant("alice"), Default::default())
            .await
            .unwrap_err();
        assert_eq!(err.code, "participant_already_joined");
        let result = join_room(
            &*store,
            &id,
            participant("alice"),
            crate::core::JoinRoomOptions {
                upsert: true,
                ..Default::default()
            },
        )
        .await
        .unwrap();
        assert_eq!(result.participants.len(), 1);
        destroy(&h, store).await;
    }

    {
        let store = fresh(&h).await;
        let id = room_id(&prefix, name);
        create_room(
            &*store,
            crate::core::CreateRoomOptions {
                room_id: Some(id.clone()),
                max_participants: Some(1),
                ..Default::default()
            },
        )
        .await
        .unwrap();
        join_room(&*store, &id, participant("alice"), Default::default())
            .await
            .unwrap();
        let err = join_room(&*store, &id, participant("bob"), Default::default())
            .await
            .unwrap_err();
        assert_eq!(err.code, "room_full");
        destroy(&h, store).await;
    }

    {
        let store = fresh(&h).await;
        let id = room_id(&prefix, name);
        create_room(
            &*store,
            crate::core::CreateRoomOptions {
                room_id: Some(id.clone()),
                ..Default::default()
            },
        )
        .await
        .unwrap();
        join_room(&*store, &id, participant("alice"), Default::default())
            .await
            .unwrap();
        let closed = close_room(&*store, &id, Default::default()).await.unwrap();
        assert_eq!(format!("{:?}", closed.state), "Closed");
        let err = join_room(&*store, &id, participant("bob"), Default::default())
            .await
            .unwrap_err();
        assert_eq!(err.code, "room_closed");
        let delivery = handle_signal(
            &*store,
            json!({"v": 1, "type": "chat", "roomId": id, "senderId": "alice", "sessionId": "session-alice",
                   "ts": 1, "seq": 1, "payload": {"text": "still here"}}),
        )
        .await
        .unwrap();
        assert_eq!(delivery.recipients.len(), 0);
        destroy(&h, store).await;
    }

    {
        let store = fresh(&h).await;
        let id = room_id(&prefix, name);
        create_room(
            &*store,
            crate::core::CreateRoomOptions {
                room_id: Some(id.clone()),
                ..Default::default()
            },
        )
        .await
        .unwrap();
        join_room(&*store, &id, participant("alice"), Default::default())
            .await
            .unwrap();
        join_room(&*store, &id, participant("bob"), Default::default())
            .await
            .unwrap();
        let result = leave_room(&*store, &id, "alice", Default::default())
            .await
            .unwrap();
        assert_eq!(
            result
                .participants
                .iter()
                .map(|p| p.participant_id.as_str())
                .collect::<Vec<_>>(),
            vec!["bob"]
        );
        assert!(store.get_participant(&id, "alice").await.unwrap().is_none());
        let err = leave_room(&*store, &id, "alice", Default::default())
            .await
            .unwrap_err();
        assert_eq!(err.code, "participant_not_found");
        destroy(&h, store).await;
    }

    // ---- signals ---------------------------------------------------------
    {
        let store = fresh(&h).await;
        let id = room_id(&prefix, name);
        create_room(
            &*store,
            crate::core::CreateRoomOptions {
                room_id: Some(id.clone()),
                ..Default::default()
            },
        )
        .await
        .unwrap();
        join_room(&*store, &id, participant("alice"), Default::default())
            .await
            .unwrap();
        let s1 = store
            .put_signal(crate::store::SignalInput {
                room_id: id.clone(),
                envelope: create_envelope("chat", &id, "alice", "s", Some(json!({"text": "one"}))),
                received_at: 1,
            })
            .await
            .unwrap();
        let s2 = store
            .put_signal(crate::store::SignalInput {
                room_id: id.clone(),
                envelope: create_envelope("chat", &id, "alice", "s", Some(json!({"text": "two"}))),
                received_at: 2,
            })
            .await
            .unwrap();
        let s3 = store
            .put_signal(crate::store::SignalInput {
                room_id: id.clone(),
                envelope: create_envelope(
                    "chat",
                    &id,
                    "alice",
                    "s",
                    Some(json!({"text": "three"})),
                ),
                received_at: 3,
            })
            .await
            .unwrap();
        assert!(s1.seq < s2.seq && s2.seq < s3.seq);
        let after = store.list_signals(&id, s2.seq).await.unwrap();
        assert_eq!(
            after.iter().map(|s| s.seq).collect::<Vec<_>>(),
            vec![s3.seq]
        );
        let all = store.list_signals(&id, 0).await.unwrap();
        assert_eq!(all.len(), 3);
        // Envelope JSON round-trips verbatim through the store.
        assert_eq!(all[1].envelope, s2.envelope);
        destroy(&h, store).await;
    }

    {
        let store = fresh(&h).await;
        let id = room_id(&prefix, name);
        create_room(
            &*store,
            crate::core::CreateRoomOptions {
                room_id: Some(id.clone()),
                ..Default::default()
            },
        )
        .await
        .unwrap();
        join_room(&*store, &id, participant("alice"), Default::default())
            .await
            .unwrap();
        join_room(&*store, &id, participant("bob"), Default::default())
            .await
            .unwrap();
        join_room(&*store, &id, participant("carol"), Default::default())
            .await
            .unwrap();

        let offer = handle_signal(
            &*store,
            json!({"v": 1, "type": "offer", "roomId": id, "senderId": "alice", "sessionId": "session-alice",
                   "targetSenderId": "carol", "ts": 1, "seq": 1, "payload": {"sdp": "v=0\r\n"}}),
        )
        .await
        .unwrap();
        assert_eq!(
            offer
                .recipients
                .iter()
                .map(|p| p.participant_id.as_str())
                .collect::<Vec<_>>(),
            vec!["carol"]
        );

        let broadcast = handle_signal(
            &*store,
            json!({"v": 1, "type": "presence", "roomId": id, "senderId": "bob", "sessionId": "session-bob",
                   "ts": 1, "seq": 1, "payload": {"state": "online"}}),
        )
        .await
        .unwrap();
        let mut got: Vec<&str> = broadcast
            .recipients
            .iter()
            .map(|p| p.participant_id.as_str())
            .collect();
        got.sort_unstable();
        assert_eq!(got, vec!["alice", "bob", "carol"]);
        destroy(&h, store).await;
    }

    {
        let store = fresh(&h).await;
        let id = room_id(&prefix, name);
        create_room(
            &*store,
            crate::core::CreateRoomOptions {
                room_id: Some(id.clone()),
                ..Default::default()
            },
        )
        .await
        .unwrap();
        join_room(&*store, &id, participant("alice"), Default::default())
            .await
            .unwrap();
        let err = handle_signal(
            &*store,
            json!({"v": 1, "type": "chat", "roomId": id, "senderId": "eve", "sessionId": "session-eve",
                   "ts": 1, "seq": 1, "payload": {"text": "hi"}}),
        )
        .await
        .unwrap_err();
        assert_eq!(err.code, "participant_not_found");
        destroy(&h, store).await;
    }

    {
        let store = fresh(&h).await;
        let id = room_id(&prefix, name);
        create_room(
            &*store,
            crate::core::CreateRoomOptions {
                room_id: Some(id.clone()),
                ..Default::default()
            },
        )
        .await
        .unwrap();
        let err = handle_signal(&*store, json!({"not": "an envelope"}))
            .await
            .unwrap_err();
        assert_eq!(err.code, "invalid_envelope");
        destroy(&h, store).await;
    }

    {
        let store = fresh(&h).await;
        let id = room_id(&prefix, name);
        create_room(
            &*store,
            crate::core::CreateRoomOptions {
                room_id: Some(id.clone()),
                ..Default::default()
            },
        )
        .await
        .unwrap();
        let join = join_room(&*store, &id, participant("alice"), Default::default())
            .await
            .unwrap();
        handle_signal(
            &*store,
            json!({"v": 1, "type": "reaction", "roomId": id, "senderId": "alice", "sessionId": "session-alice",
                   "ts": 1, "seq": 1, "payload": {"emoji": "👋"}}),
        )
        .await
        .unwrap();
        let updated = store.get_participant(&id, "alice").await.unwrap().unwrap();
        assert!(updated.last_seen_at >= join.participant.last_seen_at);
        destroy(&h, store).await;
    }

    // ---- recordings ------------------------------------------------------
    {
        let store = fresh(&h).await;
        let id = room_id(&prefix, name);
        create_room(
            &*store,
            crate::core::CreateRoomOptions {
                room_id: Some(id.clone()),
                ..Default::default()
            },
        )
        .await
        .unwrap();
        let rec = start_recording(
            &*store,
            &id,
            crate::core::StartRecordingOptions {
                metadata: Some(json!({"mime": "video/webm"})),
                ..Default::default()
            },
        )
        .await
        .unwrap();
        assert_eq!(format!("{:?}", rec.status), "Recording");
        assert_eq!(rec.room_id, id);
        let list = store.list_recordings(&id).await.unwrap();
        assert_eq!(list.len(), 1);
        assert_eq!(list[0].metadata, Some(json!({"mime": "video/webm"})));
        let fetched = store
            .get_recording(&rec.session_id)
            .await
            .unwrap()
            .expect("recording");
        assert_eq!(fetched.session_id, rec.session_id);
        let stopped = stop_recording(&*store, &rec.session_id, Default::default())
            .await
            .unwrap();
        assert_eq!(format!("{:?}", stopped.status), "Finalized");
        assert!(stopped.stopped_at.unwrap() >= stopped.started_at);
        let by_id = store.get_recording(&rec.session_id).await.unwrap().unwrap();
        assert_eq!(format!("{:?}", by_id.status), "Finalized");
        let err = stop_recording(&*store, "missing", Default::default())
            .await
            .unwrap_err();
        assert_eq!(err.code, "recording_not_found");
        destroy(&h, store).await;
    }

    // ---- state -----------------------------------------------------------
    {
        let store = fresh(&h).await;
        let id = room_id(&prefix, name);
        create_room(
            &*store,
            crate::core::CreateRoomOptions {
                room_id: Some(id.clone()),
                ..Default::default()
            },
        )
        .await
        .unwrap();
        join_room(&*store, &id, participant("alice"), Default::default())
            .await
            .unwrap();
        handle_signal(
            &*store,
            json!({"v": 1, "type": "chat", "roomId": id, "senderId": "alice", "sessionId": "session-alice",
                   "ts": 1, "seq": 1, "payload": {"text": "x"}}),
        )
        .await
        .unwrap();
        let state = get_room_state(&*store, &id).await.unwrap();
        assert_eq!(state.room.room_id, id);
        assert_eq!(state.participants.len(), 1);
        assert!(state.signal_count >= 1);
        destroy(&h, store).await;
    }

    // ---- deleteRoom (when supported) -------------------------------------
    {
        let store = fresh(&h).await;
        let id = room_id(&prefix, name);
        create_room(
            &*store,
            crate::core::CreateRoomOptions {
                room_id: Some(id.clone()),
                ..Default::default()
            },
        )
        .await
        .unwrap();
        join_room(&*store, &id, participant("alice"), Default::default())
            .await
            .unwrap();
        let rec = start_recording(&*store, &id, Default::default())
            .await
            .unwrap();
        store
            .put_signal(crate::store::SignalInput {
                room_id: id.clone(),
                envelope: create_envelope(
                    "chat",
                    &id,
                    "alice",
                    "session-alice",
                    Some(json!({"text": "x"})),
                ),
                received_at: 1,
            })
            .await
            .unwrap();
        let _ = rec;
        store.delete_room(&id).await.unwrap();
        assert!(store.get_room(&id).await.unwrap().is_none());
        assert_eq!(store.list_participants(&id).await.unwrap().len(), 0);
        assert_eq!(store.list_signals(&id, 0).await.unwrap().len(), 0);
        assert_eq!(store.list_recordings(&id).await.unwrap().len(), 0);
        assert!(store
            .get_recording(&rec.session_id)
            .await
            .unwrap()
            .is_none());
        destroy(&h, store).await;
    }
}

async fn fresh(h: &StoreHarness) -> Box<dyn Store> {
    (h.create_store)().await.expect("create_store failed")
}

async fn destroy(h: &StoreHarness, store: Box<dyn Store>) {
    if let Some(d) = &h.destroy_store {
        d(store).await;
    }
}

/// Assert helper usable from tests: check a VidcallError code.
pub fn assert_error_code(err: &VidcallError, code: &str) {
    assert_eq!(err.code, code, "unexpected error: {err:?}");
}
